### DISCUSSION TITLE
Proper support to choose what NIC to be using

### DISCUSSION THREAD

**Discussion by blixten85** [rakshasa/rtorrent#1270] [Ideas]
I would like to see a feature where i can in a proper way specify a network card or an ip-adress. (NIC preferd)

**Comment by kannibalox** [comment]
Can you provide an example of what you mean? rTorrent can already bind to a specific IP address with `network.bind_address.set`, and in general for Linux you'd normally create an IP on the NIC you want and attach bind the process's socket to that.

**Reply by blixten85** [reply]
Is that all that's required in the config if i want to bind rTorrent to wlan0?

And if you want to bind to a specific network card? For example a VPN interface?

Edit: RuTorrent just stopped responding then

**Reply by kannibalox** [reply]
As a general rule for Linux, you don't bind to devices, you bind to IP address. The default address is usually `0.0.0.0` which means "listen on all available addresses", but normally you'd get the IP address assigned to wlan0, and bind rtorrent directly to that, or to an IP address bound to the VPN's interface if you wanted limit traffic to something specific. That command is all that's required for rTorrent to bind to a specific IP, however your specific network setup may require additional work to route the traffic how you want.

ruTorrent shouldn't be effected by where `network.bind_address.set`, since it communicates with ruTorrent over the SCGI port, via either `network.scgi.open_local` or `network.scgi.open_port`.

**Reply by blixten85** [reply]
Aha ok, i can setup the VPN client in my router, and bind wlan0 ip there. 

RuTorrent seems to be reporting my external WAN ip instead of the VPN ip when doing this though, how can i make RuTorrent report the correct ip in the GUI?

**Reply by kannibalox** [reply]
If you want ruTorrent to go out through the VPN, you'll have to bind it (or the webserver running it) to the same IP as rTorrent. Unfortunately I don't know the setting for that off the top of my head, you might want to check their documentation.

If you mean that rTorrent is reporting the wrong address to trackers, you can use the `network.local_address.set` command to set the correct external IP.

**Reply by blixten85** [reply]
Ok, thank you. 

No i ment RuTorrent. But i appreciate the network.local_address.set setting :)

**Reply by blixten85** [reply]
i have tried this but it does not always seem to work

`# Set an interface with IPv4 address to bind to: [eth0|tun0]
method.insert = cfg.interface.bind, string|const|private, (cat,"wg0")
# Get IPv4 address of a given interface
method.insert = get_interface_ipv4_address, simple|private, \
    "execute.capture=bash,-c,\"$cat=\\\"echo -n \$(ip -o -4 addr show \\\",$argument.0=,\\\" | grep -Po 'inet \\\\\\\\\\K[\\\\\\\\\\d.]+')\\\"\""
# The IP address the listening socket and outgoing connections is bound to. (bind)
schedule2 = set_bind_address, 0, 0, \
    "branch=((cfg.interface.bind)), \
        ((network.bind_address.set,(get_interface_ipv4_address,(cfg.interface.bind))))"`

is this correct or are there any updated versions of this?

**Reply by Elegant996** [reply]
This changed somewhat in some version of linux that I can't recall but does require `bash`. You can use the below:

```
method.insert = get_interface_ipv4_address, simple|private, \
"execute.capture=bash, -c, \"$cat=\\\"echo -n \$(ip -o -4 addr show \\\",$argument.0=,\\\" | grep -Po '(?<=inet\\\\\\\\\\s)\\\\\\\\\\d+(\\\\\\\\\\.\\\\\\\\\\d+){3}')\\\"\""
```

If my quote counting skills are good, you should also be able to use this which is much more readible:

```
method.insert = get_interface_ipv4_address, simple|private, \
"execute.capture=bash, -c, (cat, \"echo -n \$(ip -o -4 addr show \", (argument.0), \" | grep -Po '(?<=inet\\\\s)\\\\d+(\\\\.\\\\d+){3}')\""
```

**Comment by rakshasa** [comment]
Currently working on `feature/bind-devices` which should allow you to bind to devices using `a.b.c.d%eth0` syntax.

You should be able to bind to `0.0.0.0%eth0` or `::%eth0`, and likely going to be making it possible to use `%eth0` to set the device only.

**Comment by rakshasa** [comment]
Merged support for binding to devices.
